lib/connection.cpp

```diff
@@ -746,34 +831,55 @@ void Connection::Private::consumePresenceData(Events&& presenceData)
 void Connection::Private::consumeToDeviceEvents(Events&& toDeviceEvents)
 {
 #ifdef Quotient_E2EE_ENABLED
-    // handling m.room_key to-device encrypted event
-    visitEach(toDeviceEvents, [this](const EncryptedEvent& ee) {
-        if (ee.algorithm() != OlmV1Curve25519AesSha2AlgoKey) {
-            qCDebug(E2EE) << "Encrypted event" << ee.id() << "algorithm"
-                          << ee.algorithm() << "is not supported";
-            return;
-        }
+    if (!toDeviceEvents.empty()) {
+        qCDebug(E2EE) << "Consuming" << toDeviceEvents.size() << "to-device events";
+        visitEach(toDeviceEvents, [this](const EncryptedEvent& event) {
+            if (event.algorithm() != OlmV1Curve25519AesSha2AlgoKey) {
+                qCDebug(E2EE) << "Unsupported algorithm" << event.id() << "for event" << event.algorithm();
+                return;
+            }
+            const auto decryptedEvent = sessionDecryptMessage(event);
+            if(!decryptedEvent) {
+                qCWarning(E2EE) << "Failed to decrypt event" << event.id();
+                return;
+            }
 
-        // TODO: full maintaining of the device keys
-        // with device_lists sync extention and /keys/query
-        qCDebug(E2EE) << "Getting device keys for the m.room_key sender:"
-                      << ee.senderId();
-        // encryptionManager->updateDeviceKeys();
-
-        visit(*sessionDecryptMessage(ee),
-            [this, senderKey = ee.senderKey()](const RoomKeyEvent& roomKeyEvent) {
-                if (auto* detectedRoom = q->room(roomKeyEvent.roomId()))
-                    detectedRoom->handleRoomKeyEvent(roomKeyEvent, senderKey);
-                else
-                    qCDebug(E2EE)
-                        << "Encrypted event room id" << roomKeyEvent.roomId()
-                        << "is not found at the connection" << q->objectName();
-            },
-            [](const Event& evt) {
-                qCDebug(E2EE) << "Skipping encrypted to_device event, type"
-                              << evt.matrixType();
-            });
-    });
+            visit(*decryptedEvent,
+                [this, senderKey = event.senderKey()](const RoomKeyEvent& roomKeyEvent) {
+                    if (auto* detectedRoom = q->room(roomKeyEvent.roomId())) {
+                        detectedRoom->handleRoomKeyEvent(roomKeyEvent, senderKey);
+                    } else {
+                        qCDebug(E2EE) << "Encrypted event room id" << roomKeyEvent.roomId()
+                            << "is not found at the connection" << q->objectName();
+                    }
+                },
+                [](const Event& evt) {
+                    qCDebug(E2EE) << "Skipping encrypted to_device event, type"
+                                << evt.matrixType();
+                });
+        });
+    }
+#endif
+}
+
+void Connection::Private::consumeDevicesList(DevicesList&& devicesList)
+{
+#ifdef Quotient_E2EE_ENABLED
```

Comment:
Are device lists really specific to E2EE? I mean, I understand that there's little you can do with them outside of E2EE flows but there's nothing blocking against that?

Reply:
Not really, no. i'll change it along with the move to the database